linera-execution/src/evm/revm.rs

```diff
@@ -0,0 +1,723 @@
+        } else {
+            let result = self.transact_commit_tx_data(Choice::Call, &operation)?;
+            result.output_and_logs()
+        };
+        self.write_logs(logs, "operation")?;
+        Ok(output)
+    }
+
+    fn execute_message(
+        &mut self,
+        _context: MessageContext,
+        _message: Vec<u8>,
+    ) -> Result<(), ExecutionError> {
+        // TODO(#3760): Implement execute_message for EVM
+        todo!("The execute_message part of the Ethereum smart contract has not yet been coded");
+    }
+
+    fn finalize(&mut self, _context: FinalizeContext) -> Result<(), ExecutionError> {
+        Ok(())
+    }
+}
+
+fn process_execution_result(
+    result: ExecutionResult,
+) -> Result<ExecutionResultSuccess, ExecutionError> {
+    match result {
+        ExecutionResult::Success {
+            reason,
+            gas_used: _,
+            gas_refunded: _,
+            logs,
+            output,
+        } => Ok(ExecutionResultSuccess {
+            reason,
+            logs,
+            output,
+        }),
+        ExecutionResult::Revert { gas_used, output } => {
+            let error = EvmExecutionError::Revert { gas_used, output };
+            Err(ExecutionError::EvmError(error))
+        }
+        ExecutionResult::Halt { gas_used, reason } => {
+            let error = EvmExecutionError::Halt { gas_used, reason };
+            Err(ExecutionError::EvmError(error))
+        }
+    }
+}
+
+impl<Runtime> RevmContractInstance<Runtime>
+where
+    Runtime: ContractRuntime,
+{
+    pub fn prepare(module: Vec<u8>, runtime: Runtime) -> Self {
+        let db = DatabaseRuntime::new(runtime);
+        Self { module, db }
+    }
+
+    fn transact_commit_tx_data(
+        &mut self,
+        ch: Choice,
+        vec: &[u8],
+    ) -> Result<ExecutionResultSuccess, ExecutionError> {
+        let (kind, tx_data) = match ch {
+            Choice::Create => (TxKind::Create, Bytes::copy_from_slice(vec)),
+            Choice::Call => {
+                ensure_message_length(vec.len(), 4)?;
+                forbid_execute_operation_origin(&vec[..4])?;
+                let tx_data = Bytes::copy_from_slice(vec);
+                (TxKind::Call(Address::ZERO.create(0)), tx_data)
+            }
+        };
+        let mut inspector = CallInterceptorContract {
+            db: self.db.clone(),
+        };
+        let mut evm: Evm<'_, _, _> = Evm::builder()
+            .with_ref_db(&mut self.db)
+            .with_external_context(&mut inspector)
+            .modify_tx_env(|tx| {
+                tx.clear();
+                tx.transact_to = kind;
+                tx.data = tx_data;
+            })
+            .append_handler_register(|handler| {
+                inspector_handle_register(handler);
+                let precompiles = handler.pre_execution.load_precompiles();
+                handler.pre_execution.load_precompiles = Arc::new(move || {
+                    let mut precompiles = precompiles.clone();
```

Comment:
No need to clone them.

Reply:
That seems needed.
I could not find a way to remove the clone() operation.